tensorboard/program.py

```diff
@@ -364,56 +360,77 @@ def __init__(self, wsgi_app, flags):
     max_attempts = 10 if should_scan else 1
     base_port = min(base_port + max_attempts, 0x10000) - max_attempts
 
-    # Without an explicit host, we default to serving on all interfaces,
-    # and will attempt to serve both IPv4 and IPv6 traffic through one
-    # socket.
-    self._auto_wildcard = not host
-
     for (attempt_index, port) in (
         enumerate(xrange(base_port, base_port + max_attempts))):
-      if self._auto_wildcard:
-        host = self._get_wildcard_address(port)
       try:
-        # Yes, this invokes the super initializer potentially many
-        # times. This seems to work fine, and looking at the superclass
-        # chain (type(self).__mro__) it doesn't seem that anything
-        # _should_ go wrong (nor does any superclass provide a facility
-        # to do this natively).
-        #
-        # TODO(@wchargin): Refactor so that this is no longer necessary,
-        # probably by switching composing around a Werkzeug server
-        # rather than inheriting from it.
-        super(WerkzeugServer, self).__init__(host, port, wsgi_app)
+        self._server = _TensorBoardWsgiServer(
+            wsgi_app=wsgi_app,
+            host=host,
+            port=port,
+            path_prefix=flags.path_prefix,
+        )
         break
-      except socket.error as e:
-        if hasattr(errno, 'EACCES') and e.errno == errno.EACCES:
+      except _TensorBoardFailedToBindError as e:
+        if attempt_index < max_attempts - 1:
+          continue
+        elif should_scan:
           raise TensorBoardServerException(
-              'TensorBoard must be run as superuser to bind to port %d' %
+              'TensorBoard could not bind to any port around %s '
+              '(tried %d times)'
+              % (base_port, max_attempts))
+        else:
+          raise TensorBoardServerException(e.msg)
+
+  def serve_forever(self):
+    return self._server.serve_forever()
+
+  def get_url(self):
+    return self._server.get_url()
+
+
+class _TensorBoardFailedToBindError(TensorBoardServerException):
+  pass
+
+
+class _TensorBoardWsgiServer(serving.ThreadedWSGIServer):
```

Comment:
It's probably worth a comment either here or on WerkzeugServer or both explaining what the relationship of these two "Server" classes is, since at first glance it's definitely not obvious.

Also, it seems like a subpar naming situation that WerkzeugServer is an instance of TensorBoardServer but strictly speaking contains no Werkzeug-specific logic, while _TensorBoardWsgiServer is an instance of werkzeug.serving.ThreadedWSGIServer yet not an instance of TensorBoardServer.

This could be resolved in a few ways.  One that might actually make sense is turning this class into the original WerzeugServer (by having it inherit from TensorBoardServer), and then renaming the class above to something like WerzeugScanningServer, and just changing the default server class used in the TensorBoard constructor.  (Strictly speaking the class above is more generic than that, and we could make it e.g. ScanningServer and then compose those two yet again to produce WerzeugScanningServer, but I'm not sure that's worth it.)

Reply:
> It's probably worth a comment either here or on WerkzeugServer or both
> explaining what the relationship of these two "Server" classes is,
> since at first glance it's definitely not obvious.

Yes, agreed. Will do.

> Also, it seems like a subpar naming situation that WerkzeugServer is
> an instance of TensorBoardServer but strictly speaking contains no
> Werkzeug-specific logic, while _TensorBoardWsgiServer is an instance
> of werkzeug.serving.ThreadedWSGIServer yet not an instance of
> TensorBoardServer.

Agreed, mostly. `WerkzeugServer` contains Werkzeug-specific logic in
that it delegates to `_TensorBoardWsgiServer`, which is
Werkzeug-specific.

In your estimation, what are the public API guarantees that we make
around `WerkzeugServer`? I know that `tensorboard.program` is explicitly
a public module.

(Will look at this more carefully tomorrow.)

Reply:
In practice I don't think anyone has really used WerkzeugServer; it's not the primary API we're intending to expose from `tensorboard.program`.  But at any rate I don't think that's actually an issue here since all we'd be changing in any case is the implementation/behavior, not the API per se.

I guess more broadly my point was that certainly the proposed _TensorboardWsgiServer seems more Werkzeug-specific than the  proposed WerkzeugServer so if the latter is going to maintain Werkzeug in its name, it seems like the former should too.

Another way to rework this would be to convert the scanning logic into a factory function on the original WerkzeugServer (aka current _TensorBoardWsgiServer), since that's really more like what it is than a wholly separate class.  We could then retool the `TensorBoard` class a bit so that instead of taking only a server_class it really takes a server_factory function.  If you wanted to avoid an actual API change (which I think it likely harmless) we could add it as a new keyword argument that's mutually exclusive with server_class.